server/src/sas/formatter/parser.ts

```diff
@@ -0,0 +1,304 @@
+// Copyright © 2023, SAS Institute Inc., Cary, NC, USA.  All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import type { Token as RealToken } from "../Lexer";
+import type { FoldingBlock } from "../LexerEx";
+import type { Model } from "../Model";
+import type { SyntaxProvider } from "../SyntaxProvider";
+
+interface FakeToken extends Omit<RealToken, "type"> {
+  type: "raw-data";
+}
+type Token = FakeToken | RealToken;
+
+export interface Statement {
+  type: "statement";
+  name: string;
+  children: Token[];
+  leadingComment?: Token;
+}
+
+export interface Region {
+  type: "region";
+  block?: FoldingBlock;
+  children: (Region | Statement)[];
+}
+
+interface Program {
+  type: "program";
+  children: (Region | Statement)[];
+}
+
+export type SASAST = Program | Region | Statement | Token;
+
+const isComment = (token: Token) =>
+  token.type === "comment" || token.type === "macro-comment";
+
+const getPreStatement = (parent: Program | Region) => {
+  let preStatement = parent.children[parent.children.length - 2];
```

Comment:
whether there are cases where parent.children.length is less than 2?

Reply:
Refactored to keep a reference for previous statement